documentation/decisions/adr/000n-backend-tooling.md

```diff
@@ -0,0 +1,85 @@
+# Back-end Code Quality Tools
+
+- **Status:** Proposed
+- **Last Modified:** 2023-06-30
+- **Related Issue:** [#101](https://github.com/HHS/grants-api/issues/101)
+- **Deciders:** Aaron Couch, Daphne Gold, Sammy Steiner, Gina Carson, Lucas Brown, Billy Daly
+- **Tags:** ADR
+
+## Context and Problem Statement
+
+Back-end code quality tools should facilitate and efficiently enforce linting, auto-formatting, type-checking and security concerns. They should be easily configurable to suit our specific use-cases, uniformly applying conventions while alleviating the need for individual intervention.
+
+## Decision Drivers <!-- RECOMMENDED -->
+
+- **Ease of use and configurability:** Code quality tools should be configurable to meet the specific needs of our project and enforce standards uniformly. Running the tool should be simple, and feedback should be easily implementable and understandable.
+- **Speed:** Code quality tools should be able to run in an timely manner, ideally in a pre-commit hook.
+- **Documentation and resources:** Code quality tools should have robust and helpful documentation, specifically around usage and configuration.
+- **Lean:** Selection of code quality tools should cover all use-cases with little overlap for ease of use and to minimize mental surface area required to understand the system. As in, each tool should have a specific, meaningful purpose.
+- **Enforces security:** Tooling should assist us in identifying security concerns and vulnerabilities.
+- **Well-maintained:** We have a preference towards libraries that are widely adopted and have active maintainers.
+
+## Options Considered
+
+1. Use a curated collection of Python libraries from Flask template repository (described below)
+2. Use a different or modified set of Python libraries
+
+### Dependency Management
+
+**[Poetry](https://python-poetry.org/docs/):** Python packaging and dependency management.  
+### Code Linting
+
+**[Ruff](https://beta.ruff.rs/docs/):** An extremely fast Python linter, written in Rust. Preferred for its speed and growing community adoption.[^*]
+
+### Auto-formatting
+
+**[Black](https://github.com/psf/black):** Format Python code. Compatible with [Ruff](https://beta.ruff.rs/docs/faq/#is-ruff-compatible-with-black) out of the box, however, Ruff *may* replace the need for Black at some point. See this [issue](https://github.com/astral-sh/ruff/issues/1904).
+
+### Type Checking
+
+**[Mypy](https://mypy-lang.org/):** Validate and enforce static type checking in Python.  
```

Comment:
Definitely not blocking, but when we have a minute to breathe after we get the basic repo set up with the the current tools selected, I'd like to compare mypy and pyright. 

[This comparison](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md) is probably a bit biased because it was written by the pyright team, but it highlights some important differences between the way these type checkers work:

- Strict (pyright) vs incremental (mypy) type checking
- Differences in type inference
- Speed differences (pyright claims to be faster)

Additionally, it's not highlighted explicitly in the article but pyright is connected to pylance which means that it is also the native type checker for Python in VSCode -- which may be desirable if we want to enforce or recommend certain VSCode extensions within the project

Reply:
Thanks for bringing this up, Billy! I'll add these notes to the ADR so they don't get lost.